### ISSUE TITLE
Move fava settings to source file options

### ISSUE THREAD

**Issue by corani** [aumayr/fava#214]
Tracking: [Allow unknown Options for other programs](https://bitbucket.org/blais/beancount/issues/111/allow-unknown-options-for-other-programs#)

@corani in #210
> By the way, I've been thinking, wouldn't it be nice if we would be able to put all the settings in the source file, instead of in the now separate settings file? I don't know how best to do that currently, though I've suggested @blais to open up the option directive to allow arbitrary options instead of only a predefined list.

@blais  in #210 
> It's not a bad idea to open them up, though I'd want to warn the user about unused options so we don't get into the unfortunate scenario an option is defined but has no effect. In order to do this, I'd have to build a mechanism to register for options to be at least recognized.
>
> Another problem is that the currently option values have only a limited number of types, and they're not really in the language, their semantics is parsed in the grammar. I would probably change that. In any case, you'd get at least a string.
>
> Filed a ticket: https://bitbucket.org/blais/beancount/issues/111/allow-unknown-options-for-other-programs
> 
> Can you enumerate some use cases?

**Comment by corani** [comment]
@corani in [beancount#111](https://bitbucket.org/blais/beancount/issues/111/allow-unknown-options-for-other-programs#)

Some examples of how we could use this to store settings for fava:

```
option "fava.theme" "alternative"
option "fava.editor.print-margin-column" "66"
option "fava.editor.strip-trailing-whitespace" "True"
option "fava.sidebar.show-queries" "5"
option "fava.sidebar.permalinks.Pending Transactions" "<URL>"
option "fava.sidebar.permalinks.Cleared Transactions" "<URL>"
```

Basically I propose using a "<application>.<option>" scheme. Let applications register their preamble, if necessary, they can validate the remainder themselves. Optionally, for consistency, we could move all of beancount's options to "beancount.<original-option>" and deprecate the original ones.

Some values I'd like to be able to store are: Strings, Decimals, Integers, Booleans. Lists are already possible (by using the same option key multiple times). Dictionaries could be constructed by splitting on the dots.

Conceptually, I'd like to be able to do something like this (with the options above):
```python
>>> api.getOptions("fava")
{ "theme": "alternative",
  "editor": {
    "print-margin-column": 66,
    "strip-trailing-whitespace": True
  },
  "sidebar: {
    "show-queries": 5, 
    "permalinks": {
      "Pending Transactions": "<URL>",
      "Cleared Transactions": "<URL>"
    }
  }
}
```

**Comment by aumayr** [comment]
@corani Good idea! The "division of concerns" is important here, that's why I opted for a separate settings file. But maybe integrating it into the beancount file, as you suggested, is worth a second look.

@blais suggested the following, already existing, functionality in beancount (merged with your comment on Bitbucket): 

```
plugin "fava.options" "
{ "theme": "alternative",
  "editor": {
    "print-margin-column": 66,
    "strip-trailing-whitespace": True
  },
  "sidebar: {
    "show-queries": 5, 
    "permalinks": {
      "Pending Transactions": "<URL>",
      "Cleared Transactions": "<URL>"
    }
  }
}"
```

I think this would be a good start, until @blais implements something like you suggested (custom options).

But in terms of "user experience" we have to think about how to "move existing users" over to such a new scheme. Making them change twice (from settings -> `plugin` directive -> custom options) sounds too much hassle for the normal user…

**Comment by corani** [comment]
I suggest to leave the current settings file as-is, until we have custom options. While using the `plugin` directive works, it's not particularly user-friendly for two reasons:

1. Users need to write out a valid python dict. (or write the settings as `configparser` compatible text)
2. Multiline text doesn't work well, especially when longer. Since we don't (yet) have triple-quoted strings in the beancount parser, it'll currently throw errors when you accidentally put a double-quote in the text.

**Comment by corani** [comment]
For completeness, to make the `plugin "fava.options" "{ ... }"` work, create the following:

**fava/options.py**
```python
__plugins__ = ('fava_options',)

def fava_options(entries, options_map, config=None):
    if config:
        options_map['fava'] = eval(config)
    return (entries, [])
```

This will add the fava config to the master beancount options dictionary. (`config` is a string, so instead of eval, you could pump it through `configparser` and use the appropriate format in the source file)

**Comment by blais** [comment]
I agree, to some extent.
Though I think it's nice to start using this for a little bit and get
going, and see how it goes to make sure you really do end up needing what
you think you need now. You'll know better your requirements once you use
this for a little while and then we can build the most appropriate change
to the Beancount syntax thereafter.

**Comment by aumayr** [comment]
Now that we have `custom` directives we could add the following system:

1. Parse the settings from the settings file, as we do now. 
2. Parse the settings from `custom` directives (like with `fava-sidebar-links`), and make them override 1., so we can keep the current settings file as-is, and can play with the `custom` directives. 
3. If 2. works well, convince @blais to change the Beancount syntax for `options` directives 😇

What do you think @corani @yagebu?

**Comment by corani** [comment]
Sounds like a plan 😊

On Mon, Jun 13, 2016, 23:46 Dominik Aumayr <notifications@github.com> wrote:

> Now that we have custom directives we could add the following system:
>
>    1. Parse the settings from the settings file, as we do now.
>    2. Parse the settings from custom directives (like with
>    fava-sidebar-links), and make them override 1., so we can keep the
>    current settings file as-is, and can play with the custom directives.
>    3. If 2. works well, convince @blais <https://github.com/blais> to
>    change the Beancount syntax for options directives 😇
>
> What do you think @corani <https://github.com/corani> @yagebu
> <https://github.com/yagebu>?
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/aumayr/fava/issues/214#issuecomment-225621796>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAdWB1rG2z0aS27tdNRn13pajhMrBPfTks5qLXtAgaJpZM4HzlD->
> .
>
-- 

Best regards,
Daniël Bos

Your government is reading your email. Slow them down with encryption.

My public key: http://goo.gl/gms497 (4096 bit RSA, id EF2D5D91)
Fingerprint  : D8D0 9FBE F075 F709 7B52  2F73 326C 2123 EF2D 5D91

**Comment by yagebu** [comment]
If we do this, I'd be very much in favour of removing the settings file altogether. We already use the `custom` directives for settings-like properties, so I think it's more consistent to have it all in the beancount file.

If we have two ways of specifying the options, the code dealing with it will just get too convoluted. For ~10 options, that's just not necessary.

**Comment by aumayr** [comment]
> If we do this, I'd be very much in favour of removing the settings file altogether.

I also would not be too sorry for the settings file to go, but IMHO not until blais opens up `options` to be used by fava. `custom` directives do seem like a "hack" for options.

**Comment by yagebu** [comment]
> custom directives do seem like a "hack" for options.

But then I wouldn't implement this using the `custom` directives at all. I don't see any need to "test" the concept of having the settings in the file. It's not like anything surprising would turn up, code-wide this is trivial.

If we want the settings in-file, then we should see if beancount's syntax can be extended for us (or try to be happy with just the `custom` directives).

**Comment by aumayr** [comment]
>  I don't see any need to "test" the concept of having the settings in the file. It's not like anything surprising would turn up, code-wide this is trivial.

True.

@blais, can we convince you to open up the `options`-part of Beancount? Could also be very useful for plugins!

**Comment by blais** [comment]
I need time. Am going to be away for the better part of a month. Unsure if
I'll be able to code while traveling (driving through France with no plans
other than to sample lots of great wine in generous quantities).

(A priori, this seems like just a syntactic issue; if it's just about
syntax, I think you should be able to live with the custom directive for a
little while, it's no big drama. I can open up options later on, but I want
to focus on core stuff first.)

**Comment by yagebu** [comment]
Implemented in #334 (using custom directives). This is now the only way to specify Fava's options.
